src/test/scala/examples/ConcurrentDecoupledTestingSpec.scala

```diff
@@ -0,0 +1,159 @@
+    val y = Reg(UInt(RealGCD3.num_width.W))
+    val xOriginal = Reg(UInt(RealGCD3.num_width.W))
+    val yOriginal = Reg(UInt(RealGCD3.num_width.W))
+    val p = RegInit(false.B)
+
+    val ti = RegInit(0.U(RealGCD3.num_width.W))
+    ti := ti + 1.U
+
+    decoupledInput.ready := !p
+
+    when(decoupledInput.valid && !p) {
+      x := decoupledInput.bits.a
+      y := decoupledInput.bits.b
+      xOriginal := decoupledInput.bits.a
+      yOriginal := decoupledInput.bits.b
+      p := true.B
+    }
+
+    when(p) {
+      when(x > y) {
+        x := y       // flip register values
+        y := x
+      }
+      .otherwise {
+        y := y - x
+      }
+    }
+
+    validOutput.bits.gcd := x
+    validOutput.bits.a   := xOriginal
+    validOutput.bits.b   := yOriginal
+    validOutput.valid := y === 0.U && p
+    when(validOutput.valid) {
+      p := false.B
+    }
+  }
+
+  for(i <- 0 until numberOfEngines) {
+    gcdLogic(io.input(i), io.output(i))
+  }
+}
+
+class MultiGcdCalculatorTester(c: MultiGcdCalculator) extends AdvTester(c)  {
+  private val numberOfSamples = ConcurrentDecoupledTestingSpec.numberOfSamples
+
+  private val outputHandlers = c.io.output.map { out =>
+    new ValidSink(out, (outPort: GcdOutput) => {
+      GcdResult(peek(outPort.a).toInt, peek(outPort.b).toInt, peek(outPort.gcd).toInt)
+    })
+  }
+  private val inputHandlers = c.io.input.map { in =>
+    new DecoupledSource(in, (inPorts: GcdInput, inValues: TestGCD3Values) => {
+      wire_poke(inPorts.a, inValues.a)
+      wire_poke(inPorts.b, inValues.b)
+    })
+  }
+
+  var resultCount = 0
+  var valuesInputCount: Int = 0
+
+  def addInput(driverNumber: Int, driver: DecoupledSource[GcdInput, TestGCD3Values]): Unit = {
+    val inputData = pairsToProcess.dequeue()
+    driver.inputs.enqueue(inputData)
+    valuesInputCount += 1
+     println(s"Input: $driverNumber $inputData count $valuesInputCount")
+  }
+
+  private val pairsToProcess = mutable.Queue.fill(numberOfSamples) {
+    TestGCD3Values(rnd.nextInt(numberOfSamples) + 1, rnd.nextInt(numberOfSamples) + 1)
+  }
+
+  // give everybody something to do
+  for((driver, index) <- inputHandlers.zipWithIndex) {
+    addInput(index, driver)
+  }
+
+  def checkOutput(): Unit = {
+    def ok(result: GcdResult): String = {
+      val (e, _) = GCDCalculator.computeGcdResultsAndCycles(result.x, result.y)
+      if(e == result.gcd) "ok" else s"FAILED, got ${result.gcd} should be $e"
+    }
+    outputHandlers.zipWithIndex.foreach { case (handler, index) =>
+      if(handler.outputs.nonEmpty) {
+        resultCount += 1
+        val result = handler.outputs.dequeue()
+        println(f"handler $index%3d event $resultCount%5d got output $result ${ok(result)}")
+        if (pairsToProcess.nonEmpty) {
+          addInput(index, inputHandlers(index))
+        }
+      }
+    }
+    takestep()
+  }
+
+  //noinspection LoopVariableNotUpdated
+  while(resultCount < numberOfSamples) {
+    checkOutput()
```

Comment:
This seems more convoluted than it needs to be, with the checkOutput function providing little abstraction benefit but requiring use of variables outside its scope, which can lead to confusion. Is there a better way to structure this?